FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+
+#### Call Manager
+
+The Call Manager coordinates the call stack involved in processing a given message.
+
+#### Invocation Container
+
+The Invocation Container (IC) is the environment/sandbox that runs actor code
+within the context of a single invocation. That is, reentrancy or recursion on
+the same actor will spin off another IC.
+
+The IC is an instance of a WASM engine/runtime fulfilling the FVM-side of the
+contract. The contract consists of:
+
+1. Syscalls available as importable host-provided functions.
+2. Shared types to represent data across the FVM boundary.
+3. Actor-owned memory to exchange data in syscalls.
+4. Transparent gas accounting and automatic execution halting.
+5. (Future) Dynamic linkage of popular WASM modules "blessed" by the community
+   through some governance process, in order to reduce actor bytecode size (e.g.
+   FVM SDKs, IPLD codecs, etc).
+
+Because this contract may change over time, user-deployed actors may specify an
+IC version at [`InitActor#InstallActor` time](#actor-deployment). The strategy
+for deprecating IC versions over time has not been yet determined.
+
+#### Syscalls
+
+Syscalls are functions made available to the actor as WASM imports, allowing it
+to interact with the outer environment, and to call complex operations that are
+more performant to implement in native land (e.g. cryptographic primitives).
+
+#### Kernel
+
+The Kernel is an object created by the Call Manager and injected into the
+Invocation Container. It processes syscalls, and stores state concerning the
+invocation.
+
+#### Externs
+
+Similar to syscalls being functions provided to the actor by the FVM, Externs
+are functions provided by the node to the FVM. Depending on the programming
+languages at play, Externs may need to traverse an FFI boundary.
+
+### Actors
+
+The term _Actor_ is a reference to the [actor model], a concurrent computation
+paradigm that inspires Filecoin's programming and scalability philosophies.
+
+#### Actor types
+
+We distinguish three types of actors:
+
+1. Built-in actors.
+2. Native actors.
+3. Foreign actors.
+
+![Actor execution](./fip-draft_foundational_fvm/overview.png)
+
+##### Built-in actors
+
+Built-in actors are protocol-defined actors whose WASM bytecode ships with the
+Filecoin client. The recommended approach is to embed a CAR bundle with all the
+relevant bytecode, and seeding the blockstore with it on node start.
+
+**Canonical codebase**
+
+With the definition of a universal, portable execution bytecode format, all
+Filecoin clients can rely on a single codebase for built-in actors. This speeds
+up protocol development, shortens protocol upgrade timelines, and reduces
+coordination overhead.
+
+This FIP proposes to promote the FVM built-in actors codebase (forked from the
+Forest client), to the status of **_canonical_ built-in actors**. Clients should
+adopt this codebase, but they may choose to pursue their own.
+
+##### Native actors
```

Comment:
This is going to be moderately confusing as people may assume "native" means "native code" (i.e., runs directly on the host).

Reply:
True, we need a better name for this.

Reply:
I can't think of a better name now. Let's see how much it confuses folks when it's publicly submitted.